src/frontend/screens/Settings/ProjectSettings/DeviceName/index.tsx

```diff
@@ -0,0 +1,223 @@
+    function showDiscardChangesAlert() {
+      const unsubscribe = navigation.addListener('beforeRemove', event => {
+        event.preventDefault();
+
+        const nameHasChanges = getValues('deviceName') !== deviceName;
+
+        if (nameHasChanges) {
+          Alert.alert(t(m.discardChangesAlertTitle), undefined, [
+            {style: 'cancel', text: t(m.discardChangesAlertCancelText)},
+            {
+              style: 'default',
+              text: t(m.discardChangesAlertConfirmText),
+              onPress: () => {
+                // TODO: Is going to display mode desired behavior here? Or should it go back to previous screen
+                navigation.setParams({mode: 'display'});
+              },
+            },
+          ]);
+        } else {
+          // TODO: Is going to display mode desired behavior here? Or should it go back to previous screen
+          navigation.setParams({mode: 'display'});
+        }
+      });
+
+      return () => {
+        unsubscribe();
+      };
+    },
+    [t, navigation, getValues, deviceName],
+  );
+
+  React.useEffect(
+    function updateNavigationOptions() {
+      navigation.setOptions({
+        headerRight: () => {
+          return (
+            <IconButton
+              onPress={handleSubmit(value => {
+                updateDeviceName(value.deviceName, {
+                  onSuccess: () => navigation.setParams({mode: 'display'}),
+                  onError: _err => {
+                    // TODO: Handle errors
+                  },
+                });
+              })}>
+              <SaveIcon />
+            </IconButton>
+          );
+        },
+      });
+    },
+    [handleSubmit, navigation, updateDeviceName],
+  );
+
+  return (
+    <FieldRow label={t(m.deviceNameEditHeader)}>
+      <HookFormTextInput
+        control={control}
+        name="deviceName"
+        rules={{maxLength: 100, required: true, minLength: 1}}
+        // TODO: Update HookFormTextInput implementation so that either:
+        // - the implementation fully determines the text input's base style and this component doesn't allow custom styling
+        // - this style prop is properly merged with the text input's base style in the implementation
+        style={{flex: 1, color: BLACK, fontSize: 16}}
+        showCharacterCount
+        autoFocus
+      />
+    </FieldRow>
+  );
+}
+
+export function createNavigationOptions({
+  intl,
+}: {
+  intl: (title: MessageDescriptor) => string;
+}) {
+  return ({
+    navigation,
+    route,
+  }: NativeRootNavigationProps<'DeviceName'>): NativeStackNavigationOptions => {
+    return {
+      headerTitle: intl(m.title),
+      headerRight: () =>
+        route.params.mode === 'display' ? (
+          <IconButton onPress={() => navigation.setParams({mode: 'edit'})}>
+            <EditIcon />
+          </IconButton>
+        ) : (
+          <IconButton onPress={() => {}}>
+            <SaveIcon />
+          </IconButton>
+        ),
```

Comment:
although the screen calls `navigation.setOptions()` internally to override this, it's useful+recommended to provide a value here as stated by the [React Navigation docs](https://reactnavigation.org/docs/navigation-prop/#setoptions):

> When using navigation.setOptions, we recommend specifying a placeholder in the screen's options prop and update it using navigation.setOptions. This makes sure that the delay for updating the options isn't noticeable to the user. It also makes it work with lazy-loaded screens.


Basically it serves as the placeholder/default for when you initially navigate to this screen (either from within the app or a deeplink)